### PR TITLE
Implemented recursive caching of dependency tree for SM extraction

### PR DESCRIPTION
**Pull request by ndegwamartin**
Support for loading all dependent Structure Maps during SM extraction.

### CODE REVIEW

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModel.kt [hunk 1]
```diff
@@ -489,8 +490,8 @@ constructor(
                 context = context,
                 transformSupportServices = transformSupportServices,
                 structureMapProvider = { structureMapUrl: String?, _: IWorkerContext ->
-                  structureMapUrl?.substringAfterLast("/")?.let {
-                    defaultRepository.loadResource(it)
+                  structureMapUrl?.extractLogicalIdUuid()?.let { structureMapId ->
```
Comment (owais-vd) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModel.kt:493]: `extractLogicalIdUuid()` returning an empty string but previous extension `substringAfterLast("/")` working fine. We can revert this change

File: android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModel.kt [hunk 2]
```diff
@@ -820,6 +821,28 @@ constructor(
...
+   */
+  private suspend fun loadStructureMapDependencyTree(
+    structureMapId: String,
+  ): StructureMap? {
+    val structureMap = defaultRepository.loadResource<StructureMap>(structureMapId)
+    structureMap?.also { structureMapIterator ->
+      simpleWorkerContext.cacheResource(structureMapIterator)
+
+      structureMapIterator.import?.let { canonicalTypes ->
+        canonicalTypes.forEach { loadStructureMapDependencyTree(it.value) }
```
Comment (owais-vd) [inline android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModel.kt:840]: we need to filter out the id here

```suggestion
        canonicalTypes.forEach { structureMapUrl ->
          structureMapUrl.value?.substringAfterLast("/")?.let { structureMapId ->
            loadStructureMapDependencyTree(structureMapId)
          }
        }
```

**Comment by owais-vd** [comment]
@ndegwamartin this `SimpleWorkerContext` will not work even we load the SM imports in this because this is coming from [CoreModule](https://github.com/opensrp/fhircore/blob/a3de32e3f556a5bba8a41061a5e0b6429a53ec22/android/engine/src/main/java/org/smartregister/fhircore/engine/di/CoreModule.kt#L43) dependency and the [ResourceMapper](https://github.com/google/android-fhir/blob/4169a8dc7ebaf4e6d471c2cbe10c2189cda6e3d2/datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt#L202-L205) is using a different [SimpleWorkerContext](https://github.com/google/android-fhir/blob/4169a8dc7ebaf4e6d471c2cbe10c2189cda6e3d2/datacapture/src/main/java/com/google/android/fhir/datacapture/DataCaptureConfig.kt#L74)which is coming from SDK DataCapture configuration
https://github.com/opensrp/fhircore/blob/a3de32e3f556a5bba8a41061a5e0b6429a53ec22/android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireViewModel.kt#L106

cc: @dubdabasoduba @maimoonak

**Comment by owais-vd** [comment]
@ndegwamartin i've made some changes in finding worker now i'm able to load the imported structure map and also load the structure-definition as well. The extraction started working but got another exception during extraction.

```java
org.hl7.fhir.exceptions.FHIRException: Cannot set property patient on null - value is not a primitive type (Reference) or an ElementModel type
    at org.hl7.fhir.r4.elementmodel.Element.setProperty(Element.java:378)
    at org.hl7.fhir.r4.utils.StructureMapUtilities.processTarget(StructureMapUtilities.java:1844)
    at org.hl7.fhir.r4.utils.StructureMapUtilities.executeRule(StructureMapUtilities.java:1429)
    at org.hl7.fhir.r4.utils.StructureMapUtilities.executeGroup(StructureMapUtilities.java:1416)
    at org.hl7.fhir.r4.utils.StructureMapUtilities.executeGroup(StructureMapUtilities.java:1412)
    at org.hl7.fhir.r4.utils.StructureMapUtilities.executeDependency(StructureMapUtilities.java:1475)
    at org.hl7.fhir.r4.utils.StructureMapUtilities.executeRule(StructureMapUtilities.java:1437)
    at org.hl7.fhir.r4.utils.StructureMapUtilities.executeRule(StructureMapUtilities.java:1433)
    at org.hl7.fhir.r4.utils.StructureMapUtilities.executeGroup(StructureMapUtilities.java:1416)
    at org.hl7.fhir.r4.utils.StructureMapUtilities.transform(StructureMapUtilities.java:1390)
    at com.google.android.fhir.datacapture.mapping.ResourceMapper.extractByStructureMap(ResourceMapper.kt:213)
    at com.google.android.fhir.datacapture.mapping.ResourceMapper.access$extractByStructureMap(ResourceMapper.kt:83)
    at com.google.android.fhir.datacapture.mapping.ResourceMapper$extractByStructureMap$1.invokeSuspend(Unknown Source:15)
``` 

i also got this exception previously with different workflow here https://github.com/onaio/fhir-resources/issues/2786

cc: @dubdabasoduba @maimoonak

**Comment by ndegwamartin** [comment]
> 43b7c17

Not sure if it related but take note of [the CHANGE LOG here](https://github.com/opensrp/fhircore/blob/main/CHANGELOG.md#changed) and ensure that anywhere you have resource identifiers you explicitly qualify them with the resource types within the Structure Maps.

**Review by pld — APPROVED** [review]
